modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AddWriteInvokeClosure.java

```diff
@@ -0,0 +1,154 @@
+import static org.apache.ignite.internal.storage.pagememory.mv.AbstractPageMemoryMvPartitionStorage.ALWAYS_LOAD_VALUE;
+import static org.apache.ignite.internal.storage.pagememory.mv.AbstractPageMemoryMvPartitionStorage.rowBytes;
+
+import java.nio.ByteBuffer;
+import java.util.UUID;
+import org.apache.ignite.internal.pagememory.tree.BplusTree;
+import org.apache.ignite.internal.pagememory.tree.IgniteTree.InvokeClosure;
+import org.apache.ignite.internal.pagememory.tree.IgniteTree.OperationType;
+import org.apache.ignite.internal.schema.BinaryRow;
+import org.apache.ignite.internal.storage.MvPartitionStorage;
+import org.apache.ignite.internal.storage.RowId;
+import org.apache.ignite.internal.storage.StorageException;
+import org.apache.ignite.internal.storage.TxIdMismatchException;
+import org.apache.ignite.lang.IgniteInternalCheckedException;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Implementation of {@link InvokeClosure} for {@link AbstractPageMemoryMvPartitionStorage#addWrite(RowId, BinaryRow, UUID, UUID, int)}.
+ *
+ * <p>Synchronization between reading and updating the version chain occurs due to the locks (read and write) of the page of the tree on
+ * which the version chain is located.
+ *
+ * <p>Synchronization between update operations for the version chain must be external (by {@link RowId row ID}).
+ *
+ * <p>Operation may throw {@link StorageException} and {@link TxIdMismatchException} which will cause form
+ * {@link BplusTree#invoke(Object, Object, InvokeClosure)}.
+ */
+class AddWriteInvokeClosure implements InvokeClosure<VersionChain> {
+    private final RowId rowId;
+
+    private final @Nullable BinaryRow row;
+
+    private final UUID txId;
+
+    private final UUID commitTableId;
+
+    private final int commitPartitionId;
+
+    private final AbstractPageMemoryMvPartitionStorage storage;
+
+    private @Nullable VersionChain newRow;
+
+    private @Nullable BinaryRow previousUncommittedRowVersion;
+
+    private @Nullable RowVersion toRemove;
+
+    AddWriteInvokeClosure(
+            RowId rowId,
+            @Nullable BinaryRow row,
+            UUID txId,
+            UUID commitTableId,
+            int commitPartitionId,
+            AbstractPageMemoryMvPartitionStorage storage
+    ) {
+        this.rowId = rowId;
+        this.row = row;
+        this.txId = txId;
+        this.commitTableId = commitTableId;
+        this.commitPartitionId = commitPartitionId;
+        this.storage = storage;
+    }
+
+    @Override
+    public void call(@Nullable VersionChain oldRow) throws IgniteInternalCheckedException {
+        if (oldRow == null) {
+            RowVersion newVersion = insertRowVersion(row, NULL_LINK);
+
+            newRow = VersionChain.createUncommitted(rowId, txId, commitTableId, commitPartitionId, newVersion.link(), NULL_LINK);
+
+            return;
+        }
+
+        if (oldRow.isUncommitted()) {
+            throwIfChainBelongsToAnotherTx(oldRow, txId);
```

Comment:
You don't need to pass `txId` because it's a field. Or this method must be static for some reason?

Reply:
Fix it